login/service.go

```diff
@@ -337,6 +339,187 @@ func (keycloak *KeycloakOAuthProvider) Perform(ctx *app.LoginLoginContext, confi
+				"state":          state,
+				"known_referrer": knownReferrer,
+				"err":            err,
+			}, "failed to parse referrer")
+			return redirectWithErrorForAuthorize(ctx, knownReferrer, err.Error())
+		}
+
+		ctx.ResponseData.Header().Set(
+			"Location",
+			fmt.Sprintf("%s?code=%s&state=%s", referrerURL.String(), *ctx.Code, ctx.State))
+		return ctx.TemporaryRedirect()
+	}
+
+	// First time access, redirect to oauth provider
+
+	redirect := ctx.RedirectURI
+	referrer := ctx.RequestData.Header.Get("Referer")
+	if redirect == nil {
+		if referrer == "" {
+			jerrors, _ := jsonapi.ErrorToJSONAPIErrors(ctx, goa.ErrBadRequest("referer header and redirect param are both empty. At least one should be specified."))
+			return ctx.BadRequest(jerrors)
+		}
+		redirect = &referrer
+	}
+
+	// store referrer in a state reference to redirect later
+
+	log.Debug(ctx, map[string]interface{}{
+		"referrer": referrer,
+		"redirect": redirect,
+	}, "got request from!")
+
+	stateID := ctx.State
+
+	redirect, err := keycloak.saveParamsForAuthorize(ctx, *redirect)
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, err)
+	}
+
+	err = keycloak.saveReferrer(ctx, stateID, *redirect, validRedirectURL)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"state":    stateID,
+			"referrer": referrer,
+			"redirect": redirect,
+			"err":      err,
+		}, "unable to save the state")
+		return err
+	}
+
+	redirectURL := config.AuthCodeURL(stateID.String(), oauth2.AccessTypeOnline)
+
+	ctx.ResponseData.Header().Set("Location", redirectURL)
+	return ctx.TemporaryRedirect()
+}
+
+// PerformExchange performs exchange from code to token
+func (keycloak *KeycloakOAuthProvider) PerformExchange(ctx *app.ExchangeTokenContext, config oauth.OauthConfig, serviceConfig LoginServiceConfiguration) error {
+	/* Compute all the configuration urls */
+
+	witURL, err := serviceConfig.GetWITURL(ctx.RequestData)
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, autherrors.NewInternalError(ctx, err))
+	}
+
+	redirect := ctx.Payload.RedirectURI
+	code := ctx.Payload.Code
+
+	log.Debug(ctx, map[string]interface{}{
+		"code":         code,
+		"redirect_uri": redirect,
+	}, "token request received")
+
+	keycloakToken, err := config.Exchange(ctx, *code)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"code": code,
+			"err":  err,
+		}, "keycloak exchange operation failed")
+		return jsonapi.JSONErrorResponse(ctx, autherrors.NewUnauthorizedError("Code Not Valid"))
+	}
+
+	exp := keycloakToken.Expiry.String()
+	token := &app.OauthToken{
+		AccessToken:  &keycloakToken.AccessToken,
+		Expiry:       &exp,
+		RefreshToken: &keycloakToken.RefreshToken,
+		TokenType:    &keycloakToken.TokenType,
+	}
+
+	log.Debug(ctx, map[string]interface{}{
+		"code":         code,
+		"redirect_uri": redirect,
+	}, "exchanged code to access token")
+
+	identity, newUser, err := keycloak.CreateOrUpdateIdentity(ctx, keycloakToken.AccessToken, serviceConfig)
```

Comment:
I see that there is a lot of copy&paste from Perform(). Reuse the common code instead.

Reply:
I can modify perform function and break it into smaller functions so that can be used for both

Reply:
+1